runelite-client/src/main/java/net/runelite/client/ui/ClientUI.java

```diff
@@ -688,41 +698,92 @@ private void contract()
 		pluginPanel = null;
 	}
 
-	private void expandFrameBy(final int value)
+	private void expandFrameBy(int value)
 	{
 		if (isFullScreen())
 		{
 			return;
 		}
 
-		final int result = getValidatedResult(value);
+		boolean forcedWidthIncrease = false;
+		if (config.automaticResizeType() == ExpandResizeType.KEEP_WINDOW_SIZE)
+		{
+			int minimumWidth = frame.getLayout().minimumLayoutSize(frame).width;
+			int currentWidth = frame.getWidth();
+			if (minimumWidth > currentWidth)
+			{
+				forcedWidthIncrease = true;
+				value = minimumWidth - currentWidth;
+			}
+		}
 
-		if (result != -1)
+		if (forcedWidthIncrease || config.automaticResizeType() == ExpandResizeType.KEEP_GAME_SIZE)
 		{
-			frame.setSize(result, frame.getHeight());
+			int newWindowWidth = frame.getWidth() + value;
+			int newWindowX = frame.getX();
+			Rectangle screenBounds = frame.getGraphicsConfiguration().getBounds();
+			boolean isCloseToEdge = Math.abs((frame.getX() + frame.getWidth()) -
+				(screenBounds.getX() + screenBounds.getWidth())) <= SCREEN_EDGE_CLOSE_DISTANCE;
+			boolean wouldExpandThroughEdge = frame.getX() + newWindowWidth >
+				screenBounds.getX() + screenBounds.getWidth();
+			if (!isCloseToEdge && wouldExpandThroughEdge)
+			{
+				// Move the window to the edge
+				newWindowX = (int)(screenBounds.getX() + screenBounds.getWidth()) - frame.getWidth();
+			}
+			if (wouldExpandThroughEdge)
```

Comment:
This wouldExpandThroughEdge ifs should be merged into one, and !isCloseToEdge inside for better readability